processor/src/trace/virtual_bus/circuit/prover.rs

```diff
@@ -0,0 +1,485 @@
+            return Err(ProverError::MismatchingLengthsCircuitInputs);
+        }
+
+        if !num_evaluations.is_power_of_two() {
+            return Err(ProverError::InputsMustBePowerTwo);
+        }
+
+        if num_evaluations < 2 {
+            return Err(ProverError::InputsAtLeastTwo);
+        }
+
+        let num_layers = num_evaluations.ilog2() as usize;
+        let mut p_0_vec: Vec<MultiLinearPoly<E>> = Vec::with_capacity(num_layers);
+        let mut p_1_vec: Vec<MultiLinearPoly<E>> = Vec::with_capacity(num_layers);
+        let mut q_0_vec: Vec<MultiLinearPoly<E>> = Vec::with_capacity(num_layers);
+        let mut q_1_vec: Vec<MultiLinearPoly<E>> = Vec::with_capacity(num_layers);
+
+        let p_0 = MultiLinearPoly::from_evaluations(num_den[0].to_owned())
+            .map_err(|_| ProverError::FailedToGenerateML)?;
+        let p_1 = MultiLinearPoly::from_evaluations(num_den[1].to_owned())
+            .map_err(|_| ProverError::FailedToGenerateML)?;
+        let q_0 = MultiLinearPoly::from_evaluations(num_den[2].to_owned())
+            .map_err(|_| ProverError::FailedToGenerateML)?;
+        let q_1 = MultiLinearPoly::from_evaluations(num_den[3].to_owned())
+            .map_err(|_| ProverError::FailedToGenerateML)?;
+        p_0_vec.push(p_0);
+        p_1_vec.push(p_1);
+        q_0_vec.push(q_0);
+        q_1_vec.push(q_1);
+
+        for i in 0..num_layers {
+            let (output_p_0, output_p_1, output_q_0, output_q_1) =
+                FractionalSumCircuit::compute_layer(
+                    &p_0_vec[i],
+                    &p_1_vec[i],
+                    &q_0_vec[i],
+                    &q_1_vec[i],
+                )?;
+            p_0_vec.push(output_p_0);
+            p_1_vec.push(output_p_1);
+            q_0_vec.push(output_q_0);
+            q_1_vec.push(output_q_1);
+        }
+
+        Ok(FractionalSumCircuit {
+            p_0_vec,
+            p_1_vec,
+            q_0_vec,
+            q_1_vec,
+        })
+    }
+
+    /// Computes the output values of the layer given a set of input values
+    #[allow(clippy::type_complexity)]
+    fn compute_layer(
+        inp_p_0: &MultiLinearPoly<E>,
+        inp_p_1: &MultiLinearPoly<E>,
+        inp_q_0: &MultiLinearPoly<E>,
+        inp_q_1: &MultiLinearPoly<E>,
+    ) -> Result<
+        (MultiLinearPoly<E>, MultiLinearPoly<E>, MultiLinearPoly<E>, MultiLinearPoly<E>),
+        ProverError,
+    > {
+        let len = inp_q_0.num_evaluations();
+        let outp_p_0 = (0..len / 2)
+            .map(|i| inp_p_0[i] * inp_q_1[i] + inp_p_1[i] * inp_q_0[i])
+            .collect::<Vec<E>>();
+        let outp_p_1 = (len / 2..len)
+            .map(|i| inp_p_0[i] * inp_q_1[i] + inp_p_1[i] * inp_q_0[i])
+            .collect::<Vec<E>>();
+        let outp_q_0 = (0..len / 2).map(|i| inp_q_0[i] * inp_q_1[i]).collect::<Vec<E>>();
+        let outp_q_1 = (len / 2..len).map(|i| inp_q_0[i] * inp_q_1[i]).collect::<Vec<E>>();
```

Comment:
This yields a "tangled" circuit, which is to draw out and reason about. If we view each gate of the circuit as performing a sum over projective coordinates `(p0, q0) + (p1, q1)` (the interpretation in the paper)

- `out_p_0` is the vector of all `p0`s of the next layer (gates' left numerator)
- `out_p_1` is the vector of all `p1`s of the next layer (gates' right numerator)
- `out_q_0` is the vector of all `q0`s of the next layer (gates' left denom)
- `out_q_1` is the vector of all `q1`s of the next layer (gates' right denom)

However, `out_p_1` is defined to start at index `len/2`, i.e. in the next layer, the first gate's right numerator is halfway down the first layer. I would have expected `out_p_1[0]` to be made up of indices 2 and 3 from the first layer instead.

But more generally, I find keeping track of the p's and q's like this very difficult. It would seem more natural to me to have some `ProjectiveCoordinate` type, with addition defined `(p0, q0) + (p1, q1) = (p0*q1 + p1*q0, q0*q1)`, and encode the circuit as a list of `ProjectiveCoordinates`. Then, to get the next layer, we just apply the sum to neighbors: `[layer[0] + layer[1], layer[2] + layer[3], ...]`, where `layer` is a vector of all projective coordinates for a given layer.

Reply:
To describe what is happening in the code above, we can do the following instead of lines 121-126
```rust
        let mut out_p = (0..len)
            .map(|i| inp_p_0[i] * inp_q_1[i] + inp_p_1[i] * inp_q_0[i])
            .collect::<Vec<E>>();
        let outp_p_1 = out_p.split_off(len/2);
        let outp_p_0 = out_p;
```
The use of "projective coordinate" in the paper is just to give some intuition on the technique. However, and this is why the code avoids its mention or use, the numerator (or denominators) should be viewed as multi-linear extension polynomials (especially for the deeper layers). With this view, if we denote by $p(x_0, \cdots, x_{\nu - 1})$ the multi-linear extension of all of the numerators at a certain layer, then $p(x_0, \cdots, x_{\nu - 2}, 0)$ is exactly our `*_p_0`  and $p(x_0, \cdots, x_{\nu - 2}, 1)$ is our `*_p_1` vector. This translates directly to the formulas given in the paper for the sum-checks at a given layer.
If you still think that the use of projective coordinates is more intuitive, I would suggest that we create an issue for it and tackle it once we are decided on the bigger issues.

Reply:
As I understand it, this doesn't yield the same circuit as in the paper. IIUC, it should be:

```rs
let mut out_p_0 = Vec::new();
let mut out_p_1 = Vec::new();
let mut out_q_0 = Vec::new();
let mut out_q_1 = Vec::new();

for i in 0..len/2 {
  out_p_0[i] = in_p_0[2*i] * in_q_1[2*i] + in_p_1[2*i] * in_q_0[2*i];
  out_p_1[i] = in_p_0[2*i + 1] * in_q_1[2*i + 1] + in_p_1[2*i + 1] * in_q_0[2*i + 1];
  out_q_0[i] = in_q_0[2*i] * in_q_1[2*i];
  out_q_1[i] = in_q_0[2*i + 1] * in_q_1[2*i + 1];
}
```

Reply:
Upon closer inspection, I see that your implementation is consistent with how you defined the input layer [here](https://github.com/0xPolygonMiden/miden-vm/blob/d1dd2eb248faf82190305f4e18bacd723aa1cf5b/processor/src/trace/virtual_bus/circuit/prover.rs#L217). However, I would change that too, so that it corresponds to expanding the LogUp sum for the range checker bus. Specifically, the sum expression is:

$$
\sum_{x \in \\{0 , 1\\}^{\nu}} \frac{\mathrm{rc_multiplicity}(x)}{(\alpha - \mathrm{rc_value})(x)} 
+
\frac{-\mathrm{f_s}(x)}{(\alpha - \mathrm{sv_0})(x)}
+
\frac{-\mathrm{f_s}(x)}{(\alpha - \mathrm{sv_1})(x)}
+
\frac{-\mathrm{f_s}(x)}{(\alpha - \mathrm{sv_2})(x)}
+
\frac{-\mathrm{f_s}(x)}{(\alpha - \mathrm{sv_3})(x)}
+
\frac{-\mathrm{f_m}(x)}{(\alpha - \mathrm{mv_0})(x)}
+
\frac{-\mathrm{f_m}(x)}{(\alpha - \mathrm{mv_1})(x)}
+
\frac{0(x)}{1(x)}
$$

Expanded out, we get

$$
\frac{\mathrm{rc_multiplicity}(0)}{(\alpha - \mathrm{rc_value})(0)} 
+
\frac{-\mathrm{f_s}(0)}{(\alpha - \mathrm{sv_0})(0)}
+
\frac{-\mathrm{f_s}(0)}{(\alpha - \mathrm{sv_1})(0)}
+
\frac{-\mathrm{f_s}(0)}{(\alpha - \mathrm{sv_2})(0)}
+
\frac{-\mathrm{f_s}(0)}{(\alpha - \mathrm{sv_3})(0)}
+
\frac{-\mathrm{f_m}(0)}{(\alpha - \mathrm{mv_0})(0)}
+
\frac{-\mathrm{f_m}(0)}{(\alpha - \mathrm{mv_1})(0)}
+
\frac{0(0)}{1(0)}
+
\cdots
$$

Given the tree we would build with these as the leaves, I would expect our 4 numerator/denominators to be defined as:

$$
\begin{align*}
p0 &= [\mathrm{rc_multiplicity}(0), -\mathrm{f_s}(0), -\mathrm{f_s}(0), -\mathrm{f_m}(0)] \\
p1 &= [-\mathrm{f_s}(0), -\mathrm{f_s}(0), -\mathrm{f_m}(0), 0(0)]\\
q0 &= [(\alpha - \mathrm{rc_value})(0), (\alpha - \mathrm{sv_1})(0), (\alpha - \mathrm{sv_3})(0), (\alpha - \mathrm{mv_1})(0)]\\
q1 &= [(\alpha - \mathrm{sv_0})(0), (\alpha - \mathrm{sv_2})(0), (\alpha - \mathrm{mv_0})(0), 1(0)]\\
\end{align*}
$$

compared to how they are currently defined:

$$
\begin{align*}
p0 &= [ \mathrm{rc_multiplicity}(0), -\mathrm{f_s}(0), -\mathrm{f_s}(0), -\mathrm{f_s}(0)] \\
p1 &= [ -\mathrm{f_s}(0), -\mathrm{f_m}(0), -\mathrm{f_m}(0), 0(0)] \\
q0 &= [ (\alpha - \mathrm{rc_value})(0), (\alpha - \mathrm{sv_0})(0), (\alpha - \mathrm{sv_1})(0), (\alpha - \mathrm{sv_2})(0)] \\
q1 &= [ \mathrm{sv_3})(0), \mathrm{mv_0})(0), \mathrm{mv_1})(0), 1(0)] \\
\end{align*}
$$